File: runtime-go/rt/server_stream.go
// server_stream.go — Sky.Http.Server.Stream server-side streaming
// HTTP response runtime (Cycle 4 HS-Server).
//
// Mirror image of http_stream.go (Sky.Core.Http.Stream — client-side
// incremental body reads).  Where http_stream.go reads chunks FROM
// an upstream into Sky's update loop as Msgs, this file writes
// chunks FROM a Sky handler TO an HTTP client over a long-lived
// connection.
//
// The shape:
//
//   Sky side:                              Runtime side:
//   ─────────                              ─────────────
//   Stream.stream ct handler            → ServerStream_stream:
//                                         pack (ct, handler) into a
//                                         streaming SkyResponse; the
//                                         dispatcher detects the
//                                         StreamHandler sentinel and
//                                         takes the chunk-write path.
//
//   Stream.emit chunk writer            → ServerStream_emit:
//                                         resolve writer id → handle,
//                                         w.Write(chunk) + Flush().
//
//   Stream.finish writer                → ServerStream_finish:
//                                         flag the handle closed so
//                                         late `emit`s become no-ops
//                                         and the dispatcher knows
//                                         the handler is done.
//
//   Stream.withContentType ct writer    → ServerStream_withContentType:
//                                         set Content-Type if headers
//                                         haven't been flushed yet
//                                         (best-effort; no-op otherwise).
//
// Concurrency contract:
//
//   - Each streaming response gets its own serverStreamHandle with
//     a unique id (process-global atomic counter, NEVER zero).
//
//   - Handles live on serverStreamHandles (a sync.Map) for the
//     lifetime of the user's handler.  serveStreamingResponse
//     registers + sweeps under defer so an early panic or handler
//     return cleans up.
//
//   - http.ResponseWriter is single-goroutine-write by Go convention.
//     We don't take a mutex on emit() — the user's handler runs in
//     ONE goroutine (the http.Handler goroutine) and emits
//     sequentially.  If user code spawns goroutines and emits
//     concurrently, behaviour is undefined (same as raw http stdlib).
//
//   - http.Flusher: every chunk is followed by Flush().  If the
//     underlying ResponseWriter doesn't implement Flusher (rare —
//     only middleware wrapping that doesn't pass-through Flusher),
//     stream() returns Err ErrUnavailable BEFORE invoking the
//     handler so the user gets a clear error instead of buffered
//     non-streaming output.

package rt

import (
	"fmt"
	"net/http"
	"strings"
	"sync"
	"sync/atomic"
)

// ═══════════════════════════════════════════════════════════════
// serverStreamHandle — one in-flight streaming response
// ═══════════════════════════════════════════════════════════════

// serverStreamHandle owns one open http.ResponseWriter for the
// duration of a streaming response.  The user's handler emits
// chunks via the handle's id; the dispatcher cleans up on handler
// return.
type serverStreamHandle struct {
	id      int64
	w       http.ResponseWriter
	flusher http.Flusher

	// closed — set true the first time finish fires (OR the
	// dispatcher sweeps after handler return).  Subsequent emit /
	// finish / withContentType are no-ops.
	closed atomic.Bool

	// headerSent — set true on the first successful emit OR
	// explicit pre-emit header flush.  Once true, withContentType
	// is a no-op (HTTP can't undo a flushed header).
	headerSent atomic.Bool

	// mu guards the write path against concurrent emit / finish
	// from misbehaved user code.  Optional — the documented
	// contract is single-writer, but the lock prevents a runtime
	// panic if the user violates it.
	mu sync.Mutex
}

// ═══════════════════════════════════════════════════════════════
// Registry — id → handle
// ═══════════════════════════════════════════════════════════════

// serverStreamHandles maps stream id → *serverStreamHandle for the
// duration of the user's handler.  Process-global (NOT per-session)
// so a Sky.Http.Server handler — which has no session context — can
// resolve its writer id.  Tradeoff: a runaway handler that never
// returns + never finishes leaks one entry forever; the dispatcher's
// defer-sweep neutralises this for normal handler exits + panics.
var serverStreamHandles sync.Map

// serverStreamIDCounter — monotonic source of unique stream ids.
// Process-wide, atomic for lock-free allocation under concurrent
// requests.
var serverStreamIDCounter atomic.Int64

// nextServerStreamID — fresh non-zero id.  0 is reserved so a zero-
// valued StreamWriter (uninitialised model field) can't accidentally
// resolve to a live stream.
func nextServerStreamID() int64 {
	for {
		id := serverStreamIDCounter.Add(1)
		if id != 0 {
			return id
		}
	}
}

// lookupServerStream — resolve an id to the in-flight handle.
// Returns nil if the id was never registered OR was already swept.
func lookupServerStream(id int64) *serverStreamHandle {
	v, ok := serverStreamHandles.Load(id)
	if !ok {
		return nil
	}
	return v.(*serverStreamHandle)
}

// ═══════════════════════════════════════════════════════════════
// Sky-facing kernel entries
// ═══════════════════════════════════════════════════════════════

// pendingStreamHandlers carries StreamHandler closures across the
// typed-codegen boundary.  v0.15.44 made user-side `Response` a
// typed record alias (Sky_Http_Server_Response_R) — a struct shape
// with no StreamHandler field.  When TaskCoerceT[E, Response_R]
// narrows the rt.SkyResponse the user produced via
// ServerStream_stream, the StreamHandler `any` field gets silently
// dropped (target struct has no slot for it).
//
// To preserve streaming end-to-end, ServerStream_stream registers
// the closure here under a fresh token AND stamps the token into
// the response Body (as `__sky_stream:<token>`).  The listener's
// asSkyResponse detects the sentinel + restores StreamHandler from
// this map before routing to serveStreamingResponse.  The entry is
// removed on lookup so a long-lived token can't leak indefinitely.
var pendingStreamHandlers sync.Map // map[string]any (any = handler closure)
var pendingStreamTokenSeq atomic.Int64

const pendingStreamSentinelPrefix = "__sky_stream:"

func registerPendingStreamHandler(handler any) string {
	id := pendingStreamTokenSeq.Add(1)
	token := fmt.Sprintf("%d", id)
	pendingStreamHandlers.Store(token, handler)
	return token
}

func takePendingStreamHandler(token string) (any, bool) {
	v, ok := pendingStreamHandlers.LoadAndDelete(token)
	if !ok {
		return nil, false
	}
	return v, true
}

// extractPendingStreamToken returns the token + true when body
// starts with the sentinel prefix.  Otherwise empty + false.  The
// listener uses this to detect a streaming response and look up
// its handler.
func extractPendingStreamToken(body string) (string, bool) {
	if !strings.HasPrefix(body, pendingStreamSentinelPrefix) {
		return "", false
	}
	return body[len(pendingStreamSentinelPrefix):], true
}

// ServerStream_stream implements:
//
//	Sky.Http.Server.Stream.stream
//	    : String -> (StreamWriter -> Task Error ()) -> Task Error Response
//
// Returns a Task that resolves to a `Response` carrying the
// streaming sentinel.  The dispatcher in Server_listen detects
// the non-nil StreamHandler and routes to serveStreamingResponse.
//
// The user's handler closure (the `(StreamWriter -> Task Error ())`
// argument) is stashed in pendingStreamHandlers under a unique
// token; the token is encoded into Body as `__sky_stream:<token>`
// so it survives the typed-codegen Response_R boundary.  The
// listener's asSkyResponse path detects the sentinel + restores
// the closure to StreamHandler before serving.  StreamHandler is
// ALSO set on the returned SkyResponse so the legacy any-typed
// path (FFI direct return, pre-v0.15.44 codegen) still works.
func ServerStream_stream(contentType any, handler any) any {
	ct := fmt.Sprintf("%v", contentType)
	if ct == "" {
		ct = "application/octet-stream"
	}
	token := registerPendingStreamHandler(handler)
	return func() any {
		return Ok[any, any](SkyResponse{
			Status:        200,
			ContentType:   ct,
			Body:          pendingStreamSentinelPrefix + token,
			StreamHandler: handler,
		})
	}
}

// ServerStream_emit implements:
//
//	Sky.Http.Server.Stream.emit : String -> Int -> Task Error ()
//
// (The Sky-side wrapper unwraps StreamWriter to the inner Int.)
//
// Writes the chunk + flushes.  On an unknown / closed id, returns
// Ok unit — emit-after-finish is documented as a no-op so a tail
// `finish` in the handler doesn't make a final `emit` race the
// dispatcher's sweep.
func ServerStream_emit(chunkArg any, sidArg any) any {
	id := asInt64(sidArg)
	chunk := fmt.Sprintf("%v", chunkArg)
	return func() any {
		sh := lookupServerStream(id)
		if sh == nil || sh.closed.Load() {
			return Ok[any, any](skyUnit())
		}
		sh.mu.Lock()
		defer sh.mu.Unlock()
		if sh.closed.Load() {
			return Ok[any, any](skyUnit())
		}
		// Mark header sent — once the first byte goes out the
		// status / Content-Type / withContentType is sealed.
		sh.headerSent.Store(true)
		if _, err := sh.w.Write([]byte(chunk)); err != nil {
			// Client disconnect mid-stream is the common case.
			// Mark the handle closed so subsequent emits become
			// no-ops; surface as a typed error so the handler
			// can choose to abort cleanly.
			sh.closed.Store(true)
			return Err[any, any](ErrNetwork("server.stream emit: " + err.Error()))
		}
		sh.flusher.Flush()
		return Ok[any, any](skyUnit())
	}
}

// ServerStream_finish implements:
//
//	Sky.Http.Server.Stream.finish : Int -> Task Error ()
//
// Idempotent: marking the handle closed prevents further emits.
// The dispatcher's actual unregister + return-to-net/http path
// runs after the user's handler Task fully resolves; finish just
// flips the closed flag.
func ServerStream_finish(sidArg any) any {
	id := asInt64(sidArg)
	return func() any {
		sh := lookupServerStream(id)
		if sh != nil {
			sh.closed.Store(true)
		}
		return Ok[any, any](skyUnit())
	}
}

// ServerStream_withContentType implements:
//
//	Sky.Http.Server.Stream.withContentType : String -> Int -> Task Error ()
//
// Best-effort: if headers have already been flushed (any prior
// emit) this is a no-op.  Otherwise rewrites the Content-Type on
// the underlying http.Header so the eventual flush sees the new
// value.
func ServerStream_withContentType(ctArg any, sidArg any) any {
	id := asInt64(sidArg)
	ct := fmt.Sprintf("%v", ctArg)
	return func() any {
		sh := lookupServerStream(id)
		if sh == nil || sh.closed.Load() || sh.headerSent.Load() {
			return Ok[any, any](skyUnit())
		}
		sh.mu.Lock()
		defer sh.mu.Unlock()
		if sh.headerSent.Load() {
			return Ok[any, any](skyUnit())
		}
		sh.w.Header().Set("Content-Type", ct)
		return Ok[any, any](skyUnit())
	}
}

// ═══════════════════════════════════════════════════════════════
// Dispatcher integration — called from Server_listen
// ═══════════════════════════════════════════════════════════════

// serveStreamingResponse drives a streaming response end-to-end:
//
//  1. Assert http.Flusher (reject with 503 if not implemented —
//     buffered output would silently break the streaming contract).
//  2. Apply per-response Content-Type + user-supplied headers.
//  3. Apply safe-by-default security headers (parity with the
//     buffered path).  CSRF auto-injection is skipped — streaming
//     bodies aren't form-bearing HTML.
//  4. WriteHeader (Status — defaults to 200 inside ServerStream_stream).
//  5. Flush to commit the head BEFORE invoking the user's handler;
//     this is what gets the head onto the wire so the client can
//     react to status + Content-Type before any chunk arrives
//     (critical for SSE — the EventSource spec requires the
//     "text/event-stream" Content-Type to be visible before the
//     first dispatch).
//  6. Register a serverStreamHandle, invoke the handler via SkyCall
//     with the StreamWriter ADT, drive its returned Task to
//     completion (anyTaskInvoke), sweep the handle.
//
// Panics inside the handler are caught by the outer mux closure's
// defer-recover (logPanicFrame + 500); since headers are already on
// the wire the 500 would be a no-op on the response side — the
// recover still neutralises the panic so the server stays up.
func serveStreamingResponse(w http.ResponseWriter, _ *http.Request, resp SkyResponse) {
	flusher, ok := w.(http.Flusher)
	if !ok {
		// Reverse-proxy / middleware wrapping that doesn't
		// pass-through Flusher.  Reject cleanly — buffered output
		// would defeat the whole streaming contract.
		w.Header().Set("Content-Type", "text/plain; charset=utf-8")
		w.WriteHeader(http.StatusServiceUnavailable)
		fmt.Fprint(w, "Streaming response requires http.Flusher support (Sky.Http.Server.Stream)")
		return
	}

	// Apply Content-Type from the response builder first; the user-
	// supplied Headers map overrides if needed (parity with the
	// buffered path).
	if resp.ContentType != "" {
		w.Header().Set("Content-Type", resp.ContentType)
	}
	for k, v := range resp.Headers {
		w.Header().Set(k, v)
	}
	setSecurityHeaders(w.Header())

	status := resp.Status
	if status == 0 {
		status = http.StatusOK
	}
	w.WriteHeader(status)
	// Commit the head BEFORE the handler runs — the client should
	// see the response code + Content-Type immediately, even if
	// the first chunk takes seconds to compute.
	flusher.Flush()

	sh := &serverStreamHandle{
		id:      nextServerStreamID(),
		w:       w,
		flusher: flusher,
	}
	// Pre-mark headerSent: from the user's POV the headers are
	// committed (this flush wrote them).  Any withContentType
	// call after this point is correctly a no-op.
	sh.headerSent.Store(true)

	serverStreamHandles.Store(sh.id, sh)
	defer func() {
		serverStreamHandles.Delete(sh.id)
		sh.closed.Store(true)
	}()

	// Construct the Sky-side StreamWriter ADT value — single-
	// constructor `StreamWriter Int`.  Pass it to the user's
	// handler closure via SkyCall (reflect-backed; accepts any
	// callable shape that typed codegen / `func(any) any`
	// kernel-fallback may emit).
	writerADT := SkyADT{
		Tag:     0,
		SkyName: "StreamWriter",
		Fields:  []any{sh.id},
	}
	task := SkyCall(resp.StreamHandler, writerADT)
	// Drive the handler's returned Task to completion. The result
	// is ignored — the handler's effect IS the chunk-writes; if it
	// errored we've already written some chunks + headers, so we
	// can't meaningfully signal the error to the client.  The
	// dispatcher-level connection close serves as end-of-stream.
	_ = anyTaskInvoke(task)
}

File: runtime-rust/src/sky_runtime/server_stream.rs
//! Sky.Http.Server.Stream — server-side streaming HTTP responses (chunked / SSE).
//!
//! Mirror of `runtime-go/rt/server_stream.go`. Where http_stream.rs reads an
//! upstream body chunk-by-chunk, this writes a response body chunk-by-chunk to
//! the client over a long-lived connection.
//!
//! Integration with the axum server (server.rs):
//!
//!   1. `stream ct handler` stashes the (E-erased) handler closure in a global
//!      registry under a fresh token and returns a normal `ServerResponse`
//!      whose body is the sentinel `__sky_stream:<token>`. This survives the
//!      ServerResponse bridge (which has no handler field).
//!
//!   2. `to_axum_response` (server.rs) calls `serve_streaming_sentinel`. On a
//!      sentinel hit it: pops the handler, opens a bounded mpsc channel,
//!      registers the sender under a stream id, spawns the handler driving a
//!      `StreamWriter(id)`, and returns an axum response whose body streams the
//!      channel (`Body::from_stream`). Headers + status are committed when this
//!      response is returned — before the first chunk — exactly as SSE requires.
//!
//!   3. `emit chunk writer` resolves the id → sender and `send(chunk).await`s
//!      (bounded → backpressure). `finish writer` drops the sender (ends the
//!      stream). `withContentType` is a no-op once the head is committed (which
//!      it always is by the time the handler runs — set the type via `stream`).
//!
//! `StreamWriter` is bridged (runtimeOpaqueTypes) so the runtime can construct
//! it and the stdlib's `case writer of StreamWriter raw` lowers onto it.

use super::*;
use std::collections::HashMap;
use std::future::Future;
use std::pin::Pin;
use std::sync::atomic::{AtomicI64, Ordering};
use std::sync::{Arc, Mutex, OnceLock};

/// Sky.Http.Server.Stream.StreamWriter — opaque writer handle. The variant name
/// matches the Sky constructor so `case w of StreamWriter raw` lowers onto it.
#[derive(Clone, Copy, Debug)]
pub enum StreamWriter {
    StreamWriter(i64),
}

/// Handler with its Sky error type E erased: the effect IS the emits, the
/// SkyResult is discarded (parity with the Go dispatcher's `_ = task.await`).
type ErasedStreamHandler =
    Arc<dyn Fn(StreamWriter) -> Pin<Box<dyn Future<Output = ()> + Send>> + Send + Sync>;

fn pending_handlers() -> &'static Mutex<HashMap<i64, ErasedStreamHandler>> {
    static R: OnceLock<Mutex<HashMap<i64, ErasedStreamHandler>>> = OnceLock::new();
    R.get_or_init(|| Mutex::new(HashMap::new()))
}

fn stream_senders() -> &'static Mutex<HashMap<i64, tokio::sync::mpsc::Sender<String>>> {
    static R: OnceLock<Mutex<HashMap<i64, tokio::sync::mpsc::Sender<String>>>> = OnceLock::new();
    R.get_or_init(|| Mutex::new(HashMap::new()))
}

static NEXT_TOKEN: AtomicI64 = AtomicI64::new(1);
static NEXT_STREAM_ID: AtomicI64 = AtomicI64::new(1);

const SENTINEL_PREFIX: &str = "__sky_stream:";

/// Per-process random nonce woven into the streaming sentinel. The sentinel is
/// matched on the BODY of any response, so without an unguessable component an
/// app (or a relayed upstream) whose body begins `__sky_stream:<digits>` could be
/// misread as a streaming sentinel and divert control flow. The nonce is drawn
/// once from the OS-seeded `RandomState` (std-only — no extra crate dep, so this
/// module compiles in every server/live project regardless of `Uuid` usage) so
/// body-controlled content can neither forge nor collide with a real sentinel.
/// Sentinel shape: `__sky_stream:<nonce>:<token>`.
fn sentinel_nonce() -> &'static str {
    static N: OnceLock<String> = OnceLock::new();
    N.get_or_init(|| {
        use std::hash::{BuildHasher, Hasher};
        // RandomState seeds from the OS each process; hashing two fixed values
        // mixes the two independent 64-bit seeds into 128 bits of entropy.
        let rs = std::collections::hash_map::RandomState::new();
        let mut h = rs.build_hasher();
        h.write_u64(0x5359_5F73_7472_6D31);
        let a = h.finish();
        h.write_u64(0xA5A5_5A5A_F0F0_0F0F);
        let b = h.finish();
        format!("{:016x}{:016x}", a, b)
    })
}
// Bounded channel — matches the Go runtime's streamChanBuffer (16). emit's
// `send().await` blocks when full → backpressure to the producer/relay.
const STREAM_CHAN_BUFFER: usize = 16;

/// Sky.Http.Server.Stream.stream
///   : String -> (StreamWriter -> Task Error ()) -> Task Error Response
pub fn server_stream_stream<E, H>(content_type: String, handler: H) -> SkyTask<E, ServerResponse>
where
    E: Send + 'static,
    H: Fn(StreamWriter) -> SkyTask<E, ()> + Send + Sync + 'static,
{
    // Erase E: the registry can't name the project's error type. The handler's
    // returned task is driven to completion; its result is dropped.
    let erased: ErasedStreamHandler = Arc::new(move |w: StreamWriter| {
        let task = handler(w);
        Box::pin(async move {
            let _ = task.await;
        }) as Pin<Box<dyn Future<Output = ()> + Send>>
    });
    let ct = if content_type.is_empty() {
        "application/octet-stream".to_string()
    } else {
        content_type
    };
    let token = NEXT_TOKEN.fetch_add(1, Ordering::Relaxed);
    pending_handlers().lock().unwrap_or_else(|e| e.into_inner()).insert(token, erased);
    Box::pin(async move {
        SkyResult::Ok(ServerResponse {
            status: 200,
            body: format!("{}{}:{}", SENTINEL_PREFIX, sentinel_nonce(), token),
            headers: HashMap::new(),
            contentType: ct,
        })
    })
}

/// Sky.Http.Server.Stream.emit : String -> Int -> Task Error ()
/// (the stdlib unwraps StreamWriter → Int.) Sends the chunk + flushes
/// (the channel feeds an unbuffered axum body). emit-after-finish is a no-op.
pub fn server_stream_emit<E: From<String> + Send + 'static>(
    chunk: String,
    id: i64,
) -> SkyTask<E, ()> {
    Box::pin(async move {
        let sender = stream_senders().lock().unwrap_or_else(|e| e.into_inner()).get(&id).cloned();
        match sender {
            Some(tx) => match tx.send(chunk).await {
                Ok(()) => SkyResult::Ok(()),
                // Receiver dropped — client disconnected. Surface as an error so
                // a relay's forEachChunk fail-fast stops pulling the upstream.
                Err(_) => SkyResult::Err("server.stream emit: client disconnected".to_string().into()),
            },
            None => SkyResult::Ok(()),
        }
    })
}

/// Sky.Http.Server.Stream.finish : Int -> Task Error ()
/// Idempotent — drops the sender (ends the body stream). Implicit at handler
/// return; explicit when the handler wants to release the connection early.
pub fn server_stream_finish<E: From<String> + Send + 'static>(id: i64) -> SkyTask<E, ()> {
    Box::pin(async move {
        stream_senders().lock().unwrap_or_else(|e| e.into_inner()).remove(&id);
        SkyResult::Ok(())
    })
}

/// Sky.Http.Server.Stream.withContentType : String -> Int -> Task Error ()
/// Best-effort: the head is already committed by the time the handler runs in
/// this model (axum sends headers when the streaming Response is returned), so
/// this is a no-op. Set the Content-Type via the `stream` argument instead.
pub fn server_stream_with_content_type<E: From<String> + Send + 'static>(
    _ct: String,
    _id: i64,
) -> SkyTask<E, ()> {
    Box::pin(async move { SkyResult::Ok(()) })
}

/// Called from server.rs `to_axum_response`. If `r.body` carries the streaming
/// sentinel, set up the channel + spawn the handler and return the streaming
/// axum response; otherwise None (the caller falls back to the buffered path).
pub fn serve_streaming_sentinel(r: &ServerResponse) -> Option<axum::response::Response> {
    // Sentinel shape: `__sky_stream:<nonce>:<token>`. The per-process nonce must
    // match exactly, so application/relayed body content can neither forge nor
    // collide with a real pending stream. A non-match falls through to buffered.
    let rest = r.body.strip_prefix(SENTINEL_PREFIX)?;
    let token_str = rest.strip_prefix(sentinel_nonce())?.strip_prefix(':')?;
    let token: i64 = token_str.parse().ok()?;
    let handler = pending_handlers().lock().unwrap_or_else(|e| e.into_inner()).remove(&token)?;

    let (tx, rx) = tokio::sync::mpsc::channel::<String>(STREAM_CHAN_BUFFER);
    let id = loop {
        let n = NEXT_STREAM_ID.fetch_add(1, Ordering::Relaxed);
        if n != 0 {
            break n;
        }
    };
    stream_senders().lock().unwrap_or_else(|e| e.into_inner()).insert(id, tx);

    // Drive the handler in its own task; on completion drop the sender so the
    // body stream terminates even if the handler forgot to call `finish`.
    tokio::spawn(async move {
        handler(StreamWriter::StreamWriter(id)).await;
        stream_senders().lock().unwrap_or_else(|e| e.into_inner()).remove(&id);
    });

    // Receiver → byte stream. unfold yields each chunk; None ends the body when
    // every sender has dropped (finish / handler exit).
    let body_stream = futures_util::stream::unfold(rx, |mut rx| async move {
        rx.recv()
            .await
            .map(|chunk| (Ok::<String, std::io::Error>(chunk), rx))
    });

    // Clamp to the valid HTTP range before the u16 cast (parity with server.rs's
    // buffered path) so an out-of-range Sky status can't wrap/panic the cast.
    let status = axum::http::StatusCode::from_u16(r.status.clamp(100, 599) as u16)
        .unwrap_or(axum::http::StatusCode::OK);
    let mut builder = axum::http::Response::builder().status(status);
    if !r.contentType.is_empty() {
        builder = builder.header("content-type", r.contentType.clone());
    }
    // Disable proxy buffering for SSE — same hint the Sky.Live SSE path sends.
    builder = builder.header("x-accel-buffering", "no");
    builder = builder.header("cache-control", "no-cache");
    for (k, v) in &r.headers {
        builder = builder.header(k.as_str(), v.as_str());
    }
    // On builder failure — an invalid Sky-supplied content-type / header name or
    // value makes `body()` return Err — DO NOT fall through to `None`: the
    // caller's None-fallback serves the raw `__sky_stream:<nonce>:<token>`
    // sentinel verbatim to the client (leaking the per-process nonce + emitting
    // garbage). The handler is already popped and its task already spawned, so the
    // only correct outcome is a real streaming response. Emit a 500 with an empty
    // body instead of leaking the sentinel.
    match builder.body(axum::body::Body::from_stream(body_stream)) {
        Ok(resp) => Some(resp),
        Err(_) => Some(
            axum::http::Response::builder()
                .status(axum::http::StatusCode::INTERNAL_SERVER_ERROR)
                .body(axum::body::Body::empty())
                .unwrap_or_else(|_| axum::response::Response::new(axum::body::Body::empty())),
        ),
    }
}
